File: Scala/src/main/scala/AlgorithmicHeights/PAR3.scala
package AlgorithmicHeights

object PAR3 {
  import scala.annotation.tailrec

  private def convertToIntArray(line: String): Array[Int] = line.split(" ").map(_.toInt)

  private def swapElems(array: Array[Int], ix: Int, jy: Int): Unit = {
    val elem: Int = array(ix)
    array(ix) = array(jy)
    array(jy) = elem
  }

  def threeWayPartitioning(array: Array[Int], pivot: Int, n: Int): Unit = {
    @tailrec
    def loop(middleIx: Int, currentIx: Int, endIx: Int): Unit = {
      if (currentIx > endIx) ()
      else if (array(currentIx) < pivot) {
        swapElems(array, currentIx, middleIx)
        loop(middleIx + 1, currentIx + 1, endIx)
      }
      else if (array(currentIx) > pivot) {
        swapElems(array, currentIx, endIx)
        loop(middleIx, currentIx, endIx - 1)
      }
      else loop(middleIx, currentIx + 1, endIx)
    }

    loop(middleIx = 0, currentIx = 1, endIx = n - 1)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val array: Array[Int] = convertToIntArray(reader.next())
    threeWayPartitioning(array, array(0), n)
    println(array.mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/par3.ml
let convert_to_ints (line : string) : int array =
    Array.of_list @@ List.map int_of_string Str.(line |> split (regexp " "))


let three_way_partitioning (arr : int array) (pivot : int) (n : int) : unit =
    let swap i j =
        let elem = arr.(i) in
        arr.(i) <- arr.(j);
        arr.(j) <- elem in
    let rec loop middle_ix current_ix end_ix =
        if current_ix > end_ix then ()
        else if arr.(current_ix) < pivot then
            (swap current_ix middle_ix;
            loop (middle_ix + 1) (current_ix + 1) end_ix)
        else if arr.(current_ix) > pivot then
            (swap current_ix end_ix;
            loop middle_ix current_ix (end_ix - 1))
        else
            loop middle_ix (current_ix + 1) end_ix in
    loop 0 1 (n - 1)


let () =
    let n = read_int () in
    let arr = convert_to_ints (read_line ()) in
    three_way_partitioning arr arr.(0) n;
    Array.(arr |> map string_of_int |> to_list) |> String.concat " " |> print_endline
